impl/src/main/java/org/ehcache/internal/classes/ClassInstanceProvider.java

```diff
@@ -88,6 +99,20 @@ private T newInstance(K alias, ClassInstanceConfiguration<T> config) {
     }
   }
 
+  protected void releaseInstance(T instance) {
+    if(!created.contains(instance)) {
+      throw new IllegalArgumentException("Given instance of " + instance.getClass().getName() + " is not managed by this provider");
+    }
+    if(instance instanceof Closeable) {
+      try {
+        ((Closeable)instance).close();
+        created.remove(instance);
```

Comment:
I suggest inverting these two calls, so that the instance no longer is in the created collection.